Custom Grade Scales

BlobbyBob opened the pull request:
:sparkles: Adds Scales API
:sparkles: Hierarchy Levels can have default scales set
:sparkles: Adds UI to select scale and defaults for crag/sectors/areas/lines
:sparkles: Adds simple UI to create and manage scales

BlobbyBob commented:
@dorthrithil I would be glad, if you would want to try the current state of this one.

The code is not cleaned up yet, and I did not write any tests, but all features should work and I don't expect many bugs.
The known todos are:
- Labels for default scale in create crag/sector/area forms
- Some translations

dorthrithil commented:
Hi! Did some first tests, looks nice! Looking forward to add the silly "NB Skala" to my original homecrag Nahetal: https://localcrag-nahetal-kv454.ondigitalocean.app/pages/nb-skala

Some stuff I found:

- Scale form validation is only printed in console "INVALID"
- Grade brackets throw invalid error message:

![Screenshot 2024-12-20 at 22 49 46](https://github.com/user-attachments/assets/437b9e8a-9c8e-4b8d-8c46-25cd553942a0)

- Either Instance settings or region should get a scale setting, otherwise inheriting for a crag is confusing - where does it inherit the scale from? Didn't look in the code yet to figure it out
- Without lines, the skeleton loading animation of the leveled grade distribution is always showing

Nice to have:

- Checkbox setting per instance which line types are available. Many outdoor spots (or at least their topos) either have only routes or only boulders. So it will de-clutter the UI to not show selects that only have one effective option to choose from

Will do some more testing soon, probably after the bracket bug is fixed, so I can create own scales and use them.

BlobbyBob commented:
Good catches, I forgot to do a final check in the create mode of the form, the issues should be fixed now.

I renamed the "inherit" option to "not set". Maybe then it might be a bit clearer what it does, i.e., default is taken for hierarchy above or there is just no default if no value is set.

I think the nice to have is already partially covered. The line form only offers line types where a default scale is set. It would be easy to hide if only one option is available.

dorthrithil commented on client/src/app/modules/area/area-form/area-form.component.ts:
```diff
@@ -73,26 +79,54 @@ export class AreaFormComponent implements OnInit {
    * Builds the form on component initialization.
    */
   ngOnInit() {
+    const scalesPopulated = this.scalesService.getScales().pipe(map(scales => {
```
This is used in many form components, can be moved in scalesService

dorthrithil commented on client/src/app/modules/ascent/ascent-form/ascent-form.component.ts:
```diff
@@ -88,10 +87,17 @@ export class AscentFormComponent implements OnInit {
       : this.dialogConfig.data.ascent.line;
+
+    this.scalesService.getScale(this.line.type, this.line.gradeScale).subscribe({
+      next: (scale) => {
```
No need to use next if you dont need error and complete handlers

dorthrithil commented on client/src/app/modules/ascent/ascent-list/ascent-list.component.ts:
```diff
@@ -124,7 +113,14 @@ export class AscentListComponent implements OnInit {
-  ) {}
+    protected scalesService: ScalesService,
+  ) {
+    // todo hardcoded values
```
todo

dorthrithil commented on client/src/app/modules/ascent/ascent-list/ascent-list.component.ts:
```diff
@@ -106,8 +95,8 @@ export class AscentListComponent implements OnInit {
-  public minGradeValue = GRADES['FB'][2].value; // Skip project grades
-  public maxGradeValue = GRADES['FB'].at(-1).value;
+  public minGradeValue = 0; // Skip project grades
+  public maxGradeValue = 20;
```
Dont init with random value

dorthrithil commented on client/src/app/modules/ascent/project-climbed-form-title/project-climbed-form-title.component.html:
```diff
@@ -1,7 +1,7 @@
   <h3>{{ line.name }}</h3>
   <div class="grade">
-    {{ line.grade.name | transloco }}
+    {{ scalesService.gradeNameByValue(line.type, line.gradeScale, line.gradeValue) | async | translateSpecialGrades }}
```
Would maybe make sense to create a `lineGrade` pipe, so we can just write `{{.line | lineGrade }}` instead of this line all over the app

BlobbyBob replied:
With the current version of LineGradePipe, there is an issue, probably with the asynchronous binding. I'm not really sure how to solve this (in a nice way). Could you maybe take a look?

dorthrithil replied:
Setting the pipe `pure=false` fixes the issue:

```
@Pipe({
  name: 'lineGrade',
  pure: false,
})
```

dorthrithil reviewed (changes requested):
Checked the first 30 files, will continue soon

dorthrithil commented on client/src/app/modules/shared/components/grade-distribution-bar-chart/grade-distribution-bar-chart.component.html:
```diff
@@ -1,14 +1,17 @@
-      >({{ t("plusXProjects", { projects: projectCount }) }})</span
-    >
+  <div *ngFor="let chartData of chartData">
+    <h6>{{ chartData.lineType }} {{ chartData.gradeScale }}</h6><!-- TODO how should we present these labels? -->
```
todo - ill write this down on my to-test list, maybe i have a ui idea here

dorthrithil replied:
This is still an open todo in the current code, so i unresolved the conversation

dorthrithil commented on client/src/app/modules/shared/components/leveled-grade-distribution/leveled-grade-distribution.component.ts:
```diff
@@ -31,22 +47,64 @@ export class LeveledGradeDistributionComponent implements OnInit {
+            } else if (i == scale.gradeBrackets.length - 1) {
+              labels[i] = `${this.translocoService.translate(marker("leveledGradeDistributionFrom"))} ${gradeNameByValueMap[scale.gradeBrackets[i]]}`;
+            } else {
+              labels[i] = `${gradeNameByValueMap[scale.gradeBrackets[i-1] + 1]} - ${gradeNameByValueMap[scale.gradeBrackets[i]]}`; // todo here we assume that grade values are spaced by 1
```
todo

dorthrithil commented on client/src/app/modules/todo/todo-list/todo-list.component.ts:
```diff
@@ -92,8 +85,8 @@ export class TodoListComponent implements OnInit {
-  public minGradeValue = GRADES['FB'][0].value;
-  public maxGradeValue = GRADES['FB'].at(-1).value;
+  public minGradeValue = -2;
+  public maxGradeValue = 20;
```
same question with the hardcoded default

dorthrithil commented on client/src/app/modules/todo/todo-list/todo-list.component.ts:
```diff
@@ -116,7 +109,14 @@ export class TodoListComponent implements OnInit {
-  ) {}
+    protected scalesService: ScalesService,
+  ) {
+    // todo hardcoded values
```
todo

dorthrithil commented on client/src/app/services/crud/scales.service.ts:
```diff
@@ -0,0 +1,118 @@
+
+    // Prevent parallel requests to the same endpoint
+    if (OPEN_REQUESTS[lineType][name]) {
+      return OPEN_REQUESTS[lineType][name];
```
Building the cache with ReplaySubjects is a nice way to better integrate with rxjs. Probably not much to change from the current approach

BlobbyBob replied:
ReplaySubject is similar in some ways yet different and, tbh, I don't really see how it would integrate it with the current caching approach except for getting rid of some `of()` calls, but I don't think this makes the code more understandable or has a noticeable performance impact.

dorthrithil replied:
Yeah, probably also fine like it is. Bit of personal taste, I always like to use the fancy rxjs stuff. :D But it can stay as it is :)

dorthrithil reviewed (changes requested):
34 files and the testing to go

dorthrithil commented on server/src/resources/line_resources.py:
```diff
@@ -54,6 +54,10 @@ def get(self):
+        line_type = request.args.get("line_type", None, type=LineTypeEnum)
+        grade_scale = request.args.get("grade_scale", None)
+
+        # todo add filter options for linetype+gradescale
```
todo

BlobbyBob replied:
was already implemented

dorthrithil commented on server/src/resources/sector_resources.py:
```diff
@@ -60,6 +63,10 @@ def post(self, crag_slug):
 
+        valid, error = validate_default_scales(sector_data)
+        if not valid:
+            return NotFound(error), 404
```
raise

dorthrithil commented on server/src/util/scripts/build_rankings.py:
```diff
@@ -85,7 +84,7 @@ def build_rankings():
                     if not secret and line.secret:
                         continue
-                    ascent_value = get_grade_value(line.grade_name, line.grade_scale, line.type)
+                    ascent_value = line.grade_value
```
Having a single ranking is probably still fine. I think single line type instances will still be the most common case and if there are multiple line types, you'll have to make sure the grade values kind of fit. Separate rankings for line types can easily be implemented as a follow up though

dorthrithil reviewed (changes requested):
Had a look at all the files now. Will now do some more testing

dorthrithil reviewed (changes requested):
The scale setup script should be called once in a migration at the right point in time. Trying to migrate the glees instance ends up in no scales in the instance. It's also not possible to run the setup script before migrating as the relation for scales doesnt exist.

BlobbyBob commented on client/src/app/modules/line/line-list/line-list.component.ts:
```diff
@@ -94,8 +99,11 @@ export class LineListComponent implements OnInit {
+  public scaleKey: SelectItem<{lineType: LineType, gradeScale: string} | undefined>;
+
+  public minGradeValue = -2;
+  public maxGradeValue = 9999;
```
I can try initializing it with `null`. I'm just always unsure how all those PrimeNG components handle such values

dorthrithil replied:
If they break with null values, we could just not render them until the values are ready

BlobbyBob commented on client/src/app/modules/scale/scale-form/scale-form.component.ts:
```diff
@@ -0,0 +1,260 @@
+  }
+
+  gradeControls() {
+    return this.scaleForm.controls.grades as FormArray;
```
But doesn't the pipe only help in the html part?
I use this method 10 times inside the `.ts` part of the component and would need to inline the type cast every time. I would also not worry about the efficiency of a method return a fixed property and trust the JIT compilers here :wink:

dorthrithil replied:
Good points, I guess then it can stay as is :)

dorthrithil commented on client/src/app/modules/scale/scale-form/scale-form.component.ts:
```diff
@@ -0,0 +1,260 @@
+  }
+
+  deleteGrade(index: number) {
+    this.scaleForm.disable();
```
Should work without I think

BlobbyBob commented:
> The scale setup script should be called once in a migration at the right point in time. Trying to migrate the glees instance ends up in no scales in the instance. It's also not possible to run the setup script before migrating as the relation for scales doesnt exist.

Is now fixed

dorthrithil reviewed (changes requested):
![Screenshot 2025-01-25 at 12 56 33](https://github.com/user-attachments/assets/4329257b-76a6-42bf-9a0b-f7365253ab57)
There are a lot of cancelled scale requests. Any idea why this is?

dorthrithil commented on client/src/app/modules/scale/scale-form/scale-form.component.html:
```diff
@@ -0,0 +1,158 @@
+              </div>
+              <hr class="col-12"/>
+            </ng-container>
+            <p-message
```
Hmm, seems like FormArray support is not supported yet. Probably I didn't think of that back in the days when I implemented those directives etc. I'll create an issue for adding support, and then we can update validation in some follow up issue.

BlobbyBob commented:
> There are a lot of cancelled scale requests. Any idea why this is?

No, not really. I confirmed that `ScalesService.getScale` reaches the point of calling `this.http.get` exactly once per scale. However, in my dev setup I once even observed two requests fetching the same scale. Maybe its some strange behavior with `HttpClient` in conjunction to multiple subscriptions on an `Observable`?

BlobbyBob commented:
I experimented a bit and the amount of requests really depends on the current view. While you barely see (canceled) requests on the line list, there is an extraordinary amount on the ascent list. In fact, there not only many requests being canceled, but also lots of request going out (~10 reqs per scale for me). I find this really strange concerning both log statements
```ts
    console.log(`Sending request ${lineType}/${name}`);
    const req = this.http
      .get(this.api.scales.get(lineType, name))
      .pipe(
        map((x) => (console.log(`Resolving request ${lineType}/${name}`), x)),
```
are run exactly once per scale. This is even more confusing as the only reason I imagined was HttpClient firing multiple times.

dorthrithil commented on client/src/app/modules/scale/scale-form/scale-form.component.ts:
```diff
@@ -0,0 +1,252 @@
+  }
+
+  addBracket() {
+    this.gradeBracketsControls().push(this.fb.group({name: [], value: [42]}));
```
Maybe init brackets and grades with null or max + 1 instead of 42?

dorthrithil commented on client/src/app/modules/scale/scale-form/scale-form.component.ts:
```diff
@@ -0,0 +1,252 @@
+    this.gradeControls().removeAt(index);
+  }
+
+  addBracket() {
```
Adding many many brackets creates missing color and overflow issues. It would be finde to set a max bracket as a first step and then solve the overflow and colors in separate tickets.

![Screenshot 2025-01-26 at 16 45 47](https://github.com/user-attachments/assets/0def3b29-e0aa-4c75-b39e-edcbb4d64914)
![Screenshot 2025-01-26 at 16 45 53](https://github.com/user-attachments/assets/57242dbc-2421-4fd5-9f97-b294378e02cd)

BlobbyBob replied:
As its only a frontend restriction, I've only changed the scaleForm, but not the API. For the wrapping I implemented some wrapping solution and tried to make it as clean as possible, but it is not as clean as it was before.

dorthrithil replied:
This will work for now! I created #567 as a follow-up to handle the restrictions of max 8 brackets.

dorthrithil reviewed (changes requested):
- When creating a new line, we can currently select wither route or boulder. Traditional is missing, although there is a default scale for it. 
- Also, would it be possible to allow choosing the scale when adding a e.g. boulder if two boulder scales exist in the instance? In Glees we are thinking about adding a "child"-scale and then having mixed areas with children and normal lines. Currently you can only select the default areas scale

dorthrithil commented:
> > There are a lot of cancelled scale requests. Any idea why this is?
> 
> No, not really. I confirmed that `ScalesService.getScale` reaches the point of calling `this.http.get` exactly once per scale. However, in my dev setup I once even observed two requests fetching the same scale. Maybe its some strange behavior with `HttpClient` in conjunction to multiple subscriptions on an `Observable`?

Yes, I also checked that it's only called one time. Hmm.. M

> I experimented a bit and the amount of requests really depends on the current view. While you barely see (canceled) requests on the line list, there is an extraordinary amount on the ascent list. In fact, there not only many requests being canceled, but also lots of request going out (~10 reqs per scale for me). I find this really strange concerning both log statements
> 
> ```ts
>     console.log(`Sending request ${lineType}/${name}`);
>     const req = this.http
>       .get(this.api.scales.get(lineType, name))
>       .pipe(
>         map((x) => (console.log(`Resolving request ${lineType}/${name}`), x)),
> ```
> 
> are run exactly once per scale. This is even more confusing as the only reason I imagined was HttpClient firing multiple times.

Adding the `shareReplay` operator fixes the issue for me. Can you confirm this? Seems like it was some weird interaction between the way `OPEN_REQUESTS` in combination with the cold `HttpClient` observables.

```
    const req = this.http
      .get(this.api.scales.get(lineType, name))
      .pipe(
        map((payload) => Scale.deserializeFull(payload)),
        map((scale) => this.updateCache(scale)),
        map((scale) => {
          delete OPEN_REQUESTS[lineType][name];
          return scale;
        }),
        shareReplay(1)
      );
```

BlobbyBob commented:
`shareReplay(1)` also works for me, good idea.

dorthrithil commented on client/src/app/modules/todo/todo-list/todo-list.component.html:
```diff
@@ -208,7 +230,8 @@
                       <div class="ml-1 grade-name">
-                        {{ todo.line.grade.name | transloco }}
+                        {{ todo.line | lineGrade }}
+<!--                        {{ scalesService.gradeNameByValue(todo.line.type, todo.line.gradeScale, todo.line.gradeValue) | async | translateSpecialGrades }}-->
```
Delete commented line

dorthrithil reviewed (changes requested):
From the last review, maybe the comment got a bit lost:

- When creating a new line, we can currently select wither route or boulder. Traditional is missing, although there is a default scale for it.
- Also, would it be possible to allow choosing the scale when adding a e.g. boulder if two boulder scales exist in the instance? In Glees we are thinking about adding a "child"-scale and then having mixed areas with children and normal lines. Currently you can only select the default areas scale

BlobbyBob commented:
Ah yeah. I read that but forgot about it as so many new messages appeared in the feed. 

>     * When creating a new line, we can currently select wither route or boulder. Traditional is missing, although there is a default scale for it.
>     * Also, would it be possible to allow choosing the scale when adding a e.g. boulder if two boulder scales exist in the instance? In Glees we are thinking about adding a "child"-scale and then having mixed areas with children and normal lines. Currently you can only select the default areas scale

So, how it currently works is that you can only select line types when you explicitly configured the default scale for the line type for one of the hierarchy levels (if it doesn't, thats a bug). I thought about giving an option to select a scale, but this comes with some downsides. If you, for example, create a child scale for Glees, the scale will also be selectable for Kottenheim and all the other crags. For me, this outweighed the positives. Besides it complicates the already long line form even further.

A more fine-grained scale control per hierarchy level is possible, but requires quite some rework again. Not sure whether we want this in this already huge PR or just move it to a new issue for the future. IMO this is not that urgent.

dorthrithil commented:
> So, how it currently works is that you can only select line types when you explicitly configured the default scale for the line type for one of the hierarchy levels (if it doesn't, thats a bug). I thought about giving an option to select a scale, but this comes with some downsides. If you, for example, create a child scale for Glees, the scale will also be selectable for Kottenheim and all the other crags. For me, this outweighed the positives. Besides it complicates the already long line form even further.

> A more fine-grained scale control per hierarchy level is possible, but requires quite some rework again. Not sure whether we want this in this already huge PR or just move it to a new issue for the future. IMO this is not that urgent.

I think it would not be an issue to also have it selectable in Kottenheim. In the end, you as an instance owner decide, whether you want to offer an alternative scale or not. But you're right about the PR already being quite. I'll just create an issue for it to deal with it later.

BlobbyBob commented:
I've cleaned up the open stuff and implemented a simple scale change logo. I think its fine for now, but having an open issue for that is certainly good. As an instance admin I would not trust the moderators enough to always adhere to choosing the correct rules :sweat_smile:

dorthrithil reviewed (approved):
Tested the last changes, looks good! The rest we can do in follow-ups as discussed.
